src/backend/Level0/CHIPBackendLevel0.cc

```diff
@@ -850,18 +886,55 @@ void CHIPQueueLevel0::addCallback(hipStreamCallback_t Callback,
 
 ze_command_list_handle_t CHIPQueueLevel0::getCmdList() {
   if (static_cast<CHIPBackendLevel0 *>(Backend)->getUseImmCmdLists()) {
-    return ZeCmdList_;
+    logTrace("CHIPQueueLevel0::getCmdList() returning ICL {}",
+             (void *)ZeCmdListImm_);
+    return ZeCmdListImm_;
   } else {
-    ze_command_list_handle_t ZeCmdList;
-#ifdef CHIP_DUBIOUS_LOCKS
-    LOCK(Backend->DubiousLockLevel0)
-#endif
+    logTrace("CHIPQueueLevel0::getCmdList() returning RCL {}",
+             (void *)ZeCmdListImm_);
+    return ChipCtxLz_->getCmdListReg();
+  }
+}
+
+ze_command_list_handle_t CHIPContextLevel0::getCmdListReg() {
+  LOCK(CmdListMtx) // CHIPQueueLevel0::ZeCmdListRegStack_
+  CmdListsRequested_++;
+  ze_command_list_handle_t ZeCmdList;
+  if (ZeCmdListRegStack_.size()) {
+    ZeCmdList = ZeCmdListRegStack_.top();
+    ZeCmdListRegStack_.pop();
+    CmdListsReused_++;
+  } else {
+    // If the cmd list stack for this queue was empty, create a new one
+    // This cmd list will eventually return to the stack for this queue
+    // via CHIPEventLevel0::dissassociateCmdList()
+    auto ChipDevLz = static_cast<CHIPDeviceLevel0 *>(ChipDevice_);
+    ze_command_list_desc_t Desc = ChipDevLz->getCommandListComputeDesc();
     auto Status =
-        zeCommandListCreate(ZeCtx_, ZeDev_, &CommandListDesc_, &ZeCmdList);
+        zeCommandListCreate(this->ZeCtx, ChipDevLz->get(), &Desc, &ZeCmdList);
     CHIPERR_CHECK_LOG_AND_THROW(Status, ZE_RESULT_SUCCESS,
                                 hipErrorInitializationError);
-    return ZeCmdList;
+    NumCmdListsCreated_++;
   }
+
+  logDebug("{} CHIPContextLevel0::getCmdList() returning {}", (void *)this,
+           (void *)ZeCmdList);
+  return ZeCmdList;
+}
+
+void CHIPContextLevel0::returnCmdList(ze_command_list_handle_t ZeCmdList) {
+  // assert(!static_cast<CHIPBackendLevel0 *>(Backend)->getUseImmCmdLists() &&
+  //      "ICL is used - this should never be called");
+  LOCK(CmdListMtx) // CHIPQueueLevel0::ZeCmdListRegStack_
+  // TODO this check doesn't work since we use RCL for callbacks even when ICL
+  // if (NumCmdListsCreated_ < ZeCmdListRegStack_.size() + 1) {
+  //   logError("{} CHIPContextLevel0::returnCmdList({}) returning more cmd
+  //   lists "
+  //            "than created",
+  //            (void *)this, (void *)ZeCmdList);
+  //   assert(false);
+  // }
```

Comment:
This sounds like there is a need for refactoring. It seems the issue can be solved by passing `ze_command_list_handle_t`s around wrapped in a structure which carries the needed information - E.g:

```c++
struct CommandList {
  ze_command_list_handle_t Handle;
  bool IsImmediate;
};
```

Reply:
Leftover comments - we never need to associate events with immediate command lists since those are self-managed.